cmd/plugin/cmd/upgrade_plan.go

```diff
@@ -89,80 +119,329 @@ func init() {
+			NextVersion:    repo.DefaultVersion(),
+			Source:         providerFetchSource,
+			SourceType:     providerSourceType,
+		})
+	}
+
+	return upgradePlan{Contract: contract, Providers: upgradeItems}, nil
+}
+
+func getInstalledProviders(ctx context.Context, client ctrlclient.Client) ([]operatorv1.GenericProvider, string, error) {
+	// Iterate through installed providers and create a list of upgrade plans.
+	genericProviders := []operatorv1.GenericProvider{}
+
+	contract := "v1beta1"
+
+	// Get Core Providers.
+	var coreProviderList operatorv1.CoreProviderList
+
+	if err := client.List(ctx, &coreProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of core providers from the server: %w", err)
+	}
+
+	if len(coreProviderList.Items) == 1 && coreProviderList.Items[0].Status.Contract != nil {
+		contract = *coreProviderList.Items[0].Status.Contract
+	}
+
+	for i := range coreProviderList.Items {
+		genericProviders = append(genericProviders, &coreProviderList.Items[i])
+	}
+
+	// Get Bootstrap Providers.
+	var bootstrapProviderList operatorv1.BootstrapProviderList
+
+	if err := client.List(ctx, &bootstrapProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of bootstrap providers from the server: %w", err)
+	}
+
+	for i := range bootstrapProviderList.Items {
+		genericProviders = append(genericProviders, &bootstrapProviderList.Items[i])
+	}
+
+	// Get Control Plane Providers.
+	var controlPlaneProviderList operatorv1.ControlPlaneProviderList
+
+	if err := client.List(ctx, &controlPlaneProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of control plane providers from the server: %w", err)
+	}
+
+	for i := range controlPlaneProviderList.Items {
+		genericProviders = append(genericProviders, &controlPlaneProviderList.Items[i])
+	}
+
+	// Get Infrastructure Providers.
+	var infrastructureProviderList operatorv1.InfrastructureProviderList
+
+	if err := client.List(ctx, &infrastructureProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of infrastructure providers from the server: %w", err)
+	}
+
+	for i := range infrastructureProviderList.Items {
+		genericProviders = append(genericProviders, &infrastructureProviderList.Items[i])
+	}
+
+	// Get Addon Providers.
+	var addonProviderList operatorv1.AddonProviderList
+
+	if err := client.List(ctx, &addonProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of addon providers from the server: %w", err)
+	}
+
+	for i := range addonProviderList.Items {
+		genericProviders = append(genericProviders, &addonProviderList.Items[i])
+	}
+
+	// Get IPAM Providers.
+	var ipamProviderList operatorv1.IPAMProviderList
+
+	if err := client.List(ctx, &ipamProviderList); err != nil {
+		return nil, "", fmt.Errorf("cannot get a list of ipam providers from the server: %w", err)
+	}
+
+	for i := range ipamProviderList.Items {
+		genericProviders = append(genericProviders, &ipamProviderList.Items[i])
+	}
```

Comment:
Please add runtime extension providers here too

Reply:
working on this!